### PR TITLE
[Optimizer] Implement addOptimizerVariable

### PR DESCRIPTION
**Pull request by lhs8928**
[Optimizer] Implement addOptimizerVariable
    
    - Implement addOptimizerVariable as a parameter of tensordim
    - Implement requestOptimizerVariable in manager
    
**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped
    
Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -177,7 +177,7 @@ int NeuralNetwork::initialize() {
     opt->initialize();
     for (unsigned int idx = 0; idx < n_layers; ++idx) {
       auto const &lnode = model_graph.getSortedLayerNode(idx);
-      opt->addOptimizerVariable(lnode->getObject()->getWeightsRef());
+      lnode->addOptimizerVariable(opt);
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:180]: 👍

File: nntrainer/layers/layer_node.h [hunk 2]
```diff
@@ -385,16 +385,28 @@ class LayerNode : public ml::train::Layer, public GraphNode {
...
    */
   void save(std::ofstream &file) const;
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param opt Shared pointer of optimizer
+   */
+  void addOptimizerVariable(std::shared_ptr<Optimizer> opt) {
+    if (LAYER_V2) {
+      init_context.addOptimizerVariable(opt);
```
Comment (kparichay) [inline nntrainer/layers/layer_node.h:404]: huh, this statement will have no effect. RunContext has already been created by this time (at graph initialize), and adding the optimizer variable now will not affect the weight. Either you need to do this before the graph initializes, or you must directly update the weights.

File: nntrainer/layers/layer_context.h [hunk 3]
```diff
@@ -94,12 +95,16 @@ class InitLayerContext {
...
    */
-  unsigned int requestWeight(const TensorDim &dim, const WeightInitializer init,
-                             const WeightRegularizer reg, const float reg_const,
-                             std::string name, bool trainable = true) {
-    weights_spec.emplace_back(dim, init, reg, reg_const, trainable, name);
+  unsigned int requestWeight(
+    const TensorDim &dim, const WeightInitializer init,
+    const WeightRegularizer reg, const float reg_const, std::string name,
+    bool trainable = true,
+    std::vector<TensorDim> opt_vars_dim = std::vector<TensorDim>()) {
```
Comment (kparichay) [inline nntrainer/layers/layer_context.h:105]: `requestWeight` must not have opt_vars_dim for weight in its interface. It cannot be decided by the user, but only by the optimizer.

File: nntrainer/layers/layer_context.h [hunk 4]
```diff
@@ -94,12 +95,16 @@ class InitLayerContext {
...
+    weights_spec.emplace_back(dim, init, reg, reg_const, trainable, name,
+                              opt_vars_dim);
```
Comment (kparichay) [inline nntrainer/layers/layer_context.h:107]: you can simply use empty vector here.

File: nntrainer/optimizers/optimizer_devel.h [hunk 5]
```diff
@@ -125,6 +125,14 @@ class Optimizer : public ml::train::Optimizer {
    */
   virtual void checkValidation() const;
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param weight_spec weight_spec
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  virtual void addOptimizerVariable(Weight::Spec &weight_spec) = 0;
```
Comment (kparichay) [inline nntrainer/optimizers/optimizer_devel.h:134]: please add a TODO that L142 interface must be removed.
same for optimizer_impl.

File: nntrainer/layers/layer_context.h [hunk 6]
```diff
@@ -141,6 +146,16 @@ class InitLayerContext {
    */
   const std::vector<TensorSpec> &getTensorsSpec() const { return tensors_spec; }
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param opt Shared pointer of optimizer
+   */
+  void addOptimizerVariable(std::shared_ptr<Optimizer> opt) {
```
Comment (kparichay) [inline nntrainer/layers/layer_context.h:153]: lets not add new interfaces in context.

**Comment by zhoonit** [comment]
For the record, we discussed offline to change this design a bit.

1. optimizer provides a callback to calculate 'std::vector<TensorDim> callback(TensorDim)'
2. it is passed to `manager.addOptimizerVariable(callback, bool trainable_only);`
3. manager iterates over it's weight and add weight information to it's spec (inside manager.addOptimizerVariable)

+. This implies weightspec should be owned by the manager. If this is not easily eligible for now, we can directly work on the weights instead of weightspec and delay it to future case. But beware, in this case, we lose chance to further memory-optimize the weight variable.

File: nntrainer/layers/layer_context.h [hunk 7]
```diff
@@ -134,13 +134,6 @@ class InitLayerContext {
     return weights_spec;
   }
 
-  /**
-   * @brief Get the current tensors spec
-   *
-   * @return The current tensors spec
-   */
-  const std::vector<TensorSpec> &getTensorsSpec() const { return tensors_spec; }
```
Comment (kparichay) [inline nntrainer/layers/layer_context.h:142]: please done delete this. This is needed by the graph :).

**Comment by kparichay** [comment]
> +. This implies weightspec should be owned by the manager. If this is not easily eligible for now, we can directly work on the weights instead of weightspec and delay it to future case. But beware, in this case, we lose chance to further memory-optimize the weight variable.

Weights are be owned by the manager. WeightSpecs are owned by the context of the layer themselves. When weights are created, as the memory inside the weight is allocated/set by the manager, all the memory optimizations for the weight variable exists.

File: nntrainer/optimizers/optimizer_impl.h [hunk 8]
```diff
@@ -106,6 +106,16 @@ class OptimizerImpl : public Optimizer {
...
   virtual int initialize();
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param dim Dimension of tensor to be added as a optimizer variable
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  virtual std::vector<TensorDim> addOptimizerVariable(TensorDim dim) override {
+    return std::vector<TensorDim>();
```
Comment (kparichay) [inline nntrainer/optimizers/optimizer_impl.h:116]: default should behavior should return `{}`.

File: nntrainer/optimizers/plugged_optimizer.h [hunk 9]
```diff
@@ -138,6 +138,16 @@ class PluggedOptimizer : public nntrainer::Optimizer {
...
   void checkValidation() const override { optimizer_devel->checkValidation(); }
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param dim Dimension of tensor to be added as a optimizer variable
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  virtual std::vector<TensorDim> addOptimizerVariable(TensorDim dim) override {
+    return std::vector<TensorDim>();
```
Comment (kparichay) [inline nntrainer/optimizers/plugged_optimizer.h:148]: this must be `return optimizer_devel->addOptimizerVariable(dim)`

File: nntrainer/optimizers/optimizer_devel.h [hunk 10]
```diff
@@ -125,6 +125,14 @@ class Optimizer : public ml::train::Optimizer {
    */
   virtual void checkValidation() const;
 
+  /**
+   * @brief     Add extra variables per weight if the optimizer needs any.
+   * @param dim Dimension of tensor to be added as a optimizer variable
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  virtual std::vector<TensorDim> addOptimizerVariable(TensorDim dim) = 0;
```
Comment (kparichay) [inline nntrainer/optimizers/optimizer_devel.h:134]: please consider the signature `std::vector<TensorDim> addOptimizerVariable(const TensorDim &dim)`

File: nntrainer/optimizers/optimizer_devel.h (hunk 5 above)
Comment (lhs8928) [reply nntrainer/optimizers/optimizer_devel.h]: Added

File: nntrainer/layers/layer_context.h (hunk 6 above)
Comment (lhs8928) [reply nntrainer/layers/layer_context.h]: Deleted

File: nntrainer/layers/layer_context.h (hunk 7 above)
Comment (lhs8928) [reply nntrainer/layers/layer_context.h]: Restored

File: nntrainer/optimizers/plugged_optimizer.h (hunk 9 above)
Comment (lhs8928) [reply nntrainer/optimizers/plugged_optimizer.h]: Done

File: nntrainer/optimizers/optimizer_devel.h (hunk 10 above)
Comment (lhs8928) [reply nntrainer/optimizers/optimizer_devel.h]: Done

File: nntrainer/optimizers/adam.h [hunk 11]
```diff
@@ -57,8 +57,14 @@ class Adam : public OptimizerImpl {
...
+  /**
+   * @copydoc Optimizer::addOptimizerVariable(TensorDim dim)
+   */
+  std::vector<TensorDim> addOptimizerVariable(TensorDim &dim) override;
+
   /**
    * @copydoc Optimizer::addOptimizerVariable(std::vector<Weight> &params)
+   * @todo      Remove this function
    */
   void addOptimizerVariable(std::vector<Weight> &params);
```
Comment (kparichay) [inline nntrainer/optimizers/adam.h:69]: i think you can remove this function in this PR itself.

File: nntrainer/optimizers/adam.cpp [hunk 12]
```diff
@@ -26,6 +26,14 @@ const std::string Adam::type = "adam";
 
 enum AdamParams { wm, wv };
 
+std::vector<TensorDim> Adam::addOptimizerVariable(TensorDim &dim) {
```
Comment (kparichay) [inline nntrainer/optimizers/adam.cpp:29]:
```suggestion
std::vector<TensorDim> Adam::addOptimizerVariable(const TensorDim &dim) {
```

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by kparichay — APPROVED** [review]
LGTM
good job

File: nntrainer/optimizers/adam.cpp [hunk 13]
```diff
@@ -26,17 +26,12 @@ const std::string Adam::type = "adam";
...
+  opt_vars_dim.push_back(dim); /** Add wv */
 
-    // TODO: only trainable weights must be sent to optimizer
-    if (!w.getTrainable())
-      continue;
-
-    w.addOptimizerVariable(w.getDim()); /** Add wm */
-    w.addOptimizerVariable(w.getDim()); /** Add wv */
-  }
+  return opt_vars_dim;
```
Comment (zhoonit) [inline nntrainer/optimizers/adam.cpp:34]: Super minor suggestion, how about return `{dim, dim}` ?

File: nntrainer/optimizers/plugged_optimizer.h [hunk 14]
```diff
@@ -140,13 +140,14 @@ class PluggedOptimizer : public nntrainer::Optimizer {
...
    * @retval #ML_ERROR_NONE Successful.
    * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
    */
-  void addOptimizerVariable(std::vector<Weight> &params) override {
-    optimizer_devel->addOptimizerVariable(params);
-  }
+  virtual std::vector<TensorDim>
+  addOptimizerVariable(const TensorDim &dim) override {
+    return optimizer_devel->addOptimizerVariable(dim);
+  };
```
Comment (zhoonit) [inline nntrainer/optimizers/plugged_optimizer.h:150]: no semicolon at the end?

**Review by zhoonit — APPROVED** [review]
LGTM

**Comment by zhoonit** [comment]
Good job 👍

File: nntrainer/optimizers/adam.cpp [hunk 15]
```diff
@@ -26,17 +26,8 @@ const std::string Adam::type = "adam";
...
-
-    // TODO: only trainable weights must be sent to optimizer
-    if (!w.getTrainable())
-      continue;
-
-    w.addOptimizerVariable(w.getDim()); /** Add wm */
-    w.addOptimizerVariable(w.getDim()); /** Add wv */
-  }
+std::vector<TensorDim> Adam::addOptimizerVariable(const TensorDim &dim) {
+  return {dim, dim};
```
Comment (jijoongmoon) [inline nntrainer/optimizers/adam.cpp:30]: I guess the function name is not valid anymore. It returns just vector of dimensions.. So.. Could you consider change the name?

Comment (lhs8928) [reply nntrainer/optimizers/adam.cpp:30]: rename addOptimizerVariable to getOptimizerVariableDim in optimizer :)

File: nntrainer/optimizers/plugged_optimizer.h (hunk 14 above)
Comment (lhs8928) [reply nntrainer/optimizers/plugged_optimizer.h]: Deleted

File: nntrainer/optimizers/adam.cpp (hunk 13 above)
Comment (lhs8928) [reply nntrainer/optimizers/adam.cpp]: That would be much clear than this way thanks

File: nntrainer/optimizers/adam.cpp (hunk 12 above)
Comment (lhs8928) [reply nntrainer/optimizers/adam.cpp]: Nice catch thanks :)

File: nntrainer/optimizers/adam.h (hunk 11 above)
Comment (lhs8928) [reply nntrainer/optimizers/adam.h]: Removed

File: nntrainer/layers/layer_node.h (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/layer_node.h]: Ahh... That's why added optimizer variable was not affect the weight.
I need to consider directly update the weights. Thanks 👍

File: nntrainer/layers/layer_context.h (hunk 4 above)
Comment (lhs8928) [reply nntrainer/layers/layer_context.h]: Thanks for your advice 👍

File: nntrainer/layers/layer_context.h (hunk 3 above)
Comment (lhs8928) [reply nntrainer/layers/layer_context.h]: Hmm... That means I need to remove opt_vars_dim in interface. I will amend the commit based on your opinion

**Pull request merged**